packages/mcdoc/src/runtime/completer/index.ts

```diff
@@ -0,0 +1,61 @@
+import type * as core from '@spyglassmc/core'
+import { TypeDefSymbolData } from '../../binder/index.js'
+import type { McdocType, StructTypePairField } from '../../type/index.js'
+
+export function getFields(
+	typeDef: core.DeepReadonly<McdocType>,
+): { key: string; field: core.DeepReadonly<StructTypePairField> }[] {
+	// TODO: handle attributes
+	switch (typeDef.kind) {
+		case 'union':
+			return typeDef.members.flatMap(getFields)
+		case 'struct':
+			return typeDef.fields
+				.flatMap(field => {
+					if (field.kind === 'spread') return []
+					if (typeof field.key === 'string') {
+						return [{ key: field.key, field }]
+					}
+					if (field.key.kind === 'literal') {
+						return [{ key: `${field.key.value.value}`, field }]
+					}
+					return []
+				})
+		default:
+			return []
+	}
+}
+
+export function getValues(
+	typeDef: core.DeepReadonly<McdocType>,
```

Comment:
Same here, no need tho handle references, for instance.

Reply:
`getValues` is also called for an empty array, in which case the `typeDef.item` field is used, which isn't simplified yet (because the checker doesn't give any more information that that)

Reply:
Hmm, I see good point.

Would it be possible to manually call `simplify` for the completer when encountering this case? If you have a list of dispatchers for example, you would ideally be able to provide a completion for it.

I also kind of dislike the reference resolution being duplicated here, I think ideally you would make sure that this is always a simplified mcdoctype.

Reply:
That is definitely an option. I agree the duplicated logic isn't ideal.

Reply:
I've put a TODO for this, because I don't want this part of the code to use the unstable simplify function yet

Reply:
I think by now it should be pretty stable, but not very important either.